contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -2200,50 +2208,87 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
 	return res;
 }
 
-/*
- * Helper function to execute a utility command using
- * ProcessUtility(). Caller should make sure their
- * inputs are sanitized to prevent unexpected behaviour.
- */
-static void
-exec_utility_cmd_helper(char *query_str)
+static void 
+update_bbf_server_options(char *servername, int32 query_timeout, bool isInsert)
 {
-	List	   *parsetree_list;
-	Node	   *stmt;
-	PlannedStmt *wrapper;
+	Relation	bbf_servers_def_rel;
+	TupleDesc	bbf_servers_def_rel_dsc;
+	Datum		new_record[BBF_SERVERS_DEF_NUM_COLS];
+	bool		new_record_nulls[BBF_SERVERS_DEF_NUM_COLS];
+	bool		new_record_repl[BBF_SERVERS_DEF_NUM_COLS];
+	ScanKeyData		key;
+	HeapTuple		tuple, old_tuple;
+	TableScanDesc	tblscan;
+
+	bbf_servers_def_rel = table_open(get_bbf_servers_def_oid(),
+									 RowExclusiveLock);
+	bbf_servers_def_rel_dsc = RelationGetDescr(bbf_servers_def_rel);
+
+	MemSet(new_record_nulls, false, sizeof(new_record_nulls));
+	MemSet(new_record_repl, false, sizeof(new_record_repl));
+
+	new_record[Anum_bbf_servers_def_servername - 1] = CStringGetTextDatum(servername);
+	new_record[Anum_bbf_servers_def_query_timeout - 1] = Int32GetDatum(query_timeout);
+
+	if(isInsert)
+	{
+		tuple = heap_form_tuple(bbf_servers_def_rel_dsc,
+								new_record, new_record_nulls);
+		CatalogTupleInsert(bbf_servers_def_rel, tuple);
+	}
+	else
+	{
+		/* Search and obtain the tuple based on the server_id */	
+		ScanKeyInit(&key,
+					Anum_bbf_servers_def_servername,
+					BTEqualStrategyNumber, F_TEXTEQ,
+					CStringGetTextDatum(servername));
+		tblscan = table_beginscan_catalog(bbf_servers_def_rel, 1, &key);
+		old_tuple = heap_getnext(tblscan, ForwardScanDirection);
+
+		if (!old_tuple)
+			ereport(ERROR,
```

Comment:
end scan and close table before reporting error

Reply:
done